pepper-apis/dsm-core/src/main/java/org/broadinstitute/dsm/util/KitUtil.java

```diff
@@ -172,96 +173,116 @@ public static void createLabel(List<KitRequestCreateLabel> kitsLabelTriggered, E
+                    // DSM validates the address that is in the upload file and sets the address id if it is valid.
+                    // So here in time of label creation we can use the address id to get the address from EasyPost and use it
+                    // to buy the shipment.
+                    toAddress = KitRequestShipping.getToAddressId(easyPostUtil, kitToCreateLabelFor.getKitRequestSettings(),
+                            kitToCreateLabelFor.getAddressIdTo(), null, ddpInstanceDto);
+                    //uploaded pt is missing collaborator ids -> possibly due to an old kit being from a migrated legacy study
+                    if (kitToCreateLabelFor.getParticipantCollaboratorId() == null) {
+                        //checking if this part of the code ever happens
+                        logger.error("Had a kit without a collaborator participant id in need of a label, kit id {} in study {}",
+                                kitToCreateLabelFor.getDsmKitId(), kitToCreateLabelFor.getInstanceName());
+                        createCollaboratorPartiticpantId(kitToCreateLabelFor, ddpInstanceDto);
+                    }
+                }
+                if (toAddress != null) {
+                    try {
+                        buyShipmentForKit(easyPostUtil, kitToCreateLabelFor.getDsmKitId(), kitToCreateLabelFor.getKitRequestSettings(),
+                                kitToCreateLabelFor.getKitTyp(), toAddress, kitToCreateLabelFor.getBillingReference(),
+                                ddpInstanceDto);
+                    } catch (Exception e) {
+                        logger.error("Error buying shipment for kit " + kitToCreateLabelFor.getParticipantCollaboratorId(), e);
                     }
                 }
             } catch (Exception e) {
                 // note error and skip ptp
                 logger.error("Error getting address for participant {} in study {} with kit ID {}",
-                        kitLabel.getDdpParticipantId(), kitLabel.getInstanceName(), kitLabel.getDsmKitId(), e);
-            }
-            if (toAddress != null) {
-                //TODO: why not pass toAddress? - DC
-                buyShipmentForKit(easyPostUtil, kitLabel.getDsmKitId(), kitLabel.getKitRequestSettings(),
-                        kitLabel.getKitTyp(), toAddress.getId(), kitLabel.getBillingReference(), ddpInstanceDto);
+                        kitToCreateLabelFor.getDdpParticipantId(), kitToCreateLabelFor.getInstanceName(), kitToCreateLabelFor.getDsmKitId(),
+                        e);
             }
         }
-
         DBUtil.updateBookmark(0, BOOKMARK_LABEL_CREATION_RUNNING);
     }
 
-    protected static DDPParticipant getDDPParticipant(String ddpParticipantId, DDPInstance ddpInstance) {
+    private static void createCollaboratorPartiticpantId(KitRequestCreateLabel kitToCreateLabelFor, DDPInstanceDto ddpInstanceDto) {
```

Comment:
New method--can you put in some javadoc since there's a lot going on in this one?  Also might be worth breaking out a separate method to generate the id from updating the value in the database/dtos.